Fixes #31638 - Move the host status page from foreman_wreckingball to core

kamils-iRonin opened the pull request:
The idea is to show all statuses that implement the `.stats` method. Each hosts count number should be clickable and should redirect to the hosts page with the appropriate search query.

![image](https://user-images.githubusercontent.com/37402924/105057966-0a470780-5a76-11eb-9d0c-39a3c6d06273.png)

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ares commented:
Nice! I'm thinking whether we should also display states with 0 relevant hosts. I think we don't have the API for that yet, but it should be possible to find all subclasses of `HostStatus::Status`.

>  Each hosts count number should be clickable and should redirect to the hosts page with the appropriate search query.

I assume you mean it will be clickable (it's not today). I'd suggest also making the label clickable, the number is quite small to click. Or make it larger by saying "0 hosts" however it would repeat the word.

ares commented:
also cc @MariSvirik if you have any more thoughts for this page

kamils-iRonin commented:
I added information about the hosts owned by the current user
<img width="1296" alt="image" src="https://user-images.githubusercontent.com/37402924/105703811-1582b300-5f0e-11eb-841b-e702ca495234.png">

kamils-iRonin commented:
[test foreman]

timogoebel commented:
@kamils-iRonin: @ares and I chatted about the status of the PR. Are there any plans to implement this in react or pf-4 (could be a separate PR or just a ticket, let's talk!) so this fits the style of the new host page that is currently being made.

kamils-iRonin commented:
> Are there any plans to implement this in react or pf-4 (could be a separate PR or just a ticket, let's talk!) so this fits the style of the new host page that is currently being made.

 @timogoebel It should be pretty easy to refactor this to pf-4. It could even be this PR. I'm not sure if we need React, as this is a quite simple page.

MariSvirik commented:
I am all for changing it to PF4. In fact, it would make more sense to me, taking into account ongoing work on the new Host detail page. These two pages should be definitely consistent with each other and preferably using the same design system. Then I will be more helpful in giving you design feedback as I am more familiar with PF4 and designing the Host detail page. @kamils-iRonin feel free to ping me anytime.

kamils-iRonin commented:
@MariSvirik WDYT?

![image](https://user-images.githubusercontent.com/37402924/106654116-fcc26f00-6597-11eb-9bb4-95afebfb2e30.png)

MariSvirik commented:
@kamils-iRonin Usage of PF4 has a huge and positive impact. Personally, I really like it. But I need to ask my colleagues about the usage of the [alert component](https://www.patternfly.org/v4/components/alert/design-guidelines) in this use case, normally we don't use it this way.

There are just a couple of other things such as the size of the secondary status icons; selection of the icons next to the numbers (pficon arrows). [PF4 used to use this icon in two meanings, but that shifted a bit (see in screenshot)](https://www.patternfly.org/v4/guidelines/icons/#all-icons): 

<img width="833" alt="Screenshot 2021-02-03 at 11 40 22" src="https://user-images.githubusercontent.com/51960547/106735514-a21b2880-6614-11eb-985b-dcd76dece992.png">

<img width="830" alt="Screenshot 2021-02-03 at 11 37 41" src="https://user-images.githubusercontent.com/51960547/106735336-641e0480-6614-11eb-81ed-300ed9670436.png">

We tend to use more often an external icon but just in case it navigates to an external page (not within product/ bundle e.g. to Documentation). 
<img width="825" alt="Screenshot 2021-02-03 at 12 06 22" src="https://user-images.githubusercontent.com/51960547/106738484-42bf1780-6618-11eb-8902-eb16569fb9df.png">

Personally, I would get rid of all those icons. It looks pretty busy with them. 
 If you want we can discuss the interaction and other things such as content in the expanded cards in more detail in chat or other channels if u prefer (I also believe we can find a solution that wouldn't need an expansion of the card.). We can even schedule a meeting if u are interested and go through the designs.

kamils-iRonin commented on package.json:
```diff
@@ -21,7 +21,7 @@
   },
   "dependencies": {
-    "@theforeman/vendor": "^6.0.0",
+    "@theforeman/vendor": "^8.2.0",
```
Is there any reason to use the old `@theforeman/vendor` version? It took me a while to figure out why some PF4 features don't work

evgeni replied:
I think the 7.x branch was somehow broken, but 8.x should be fine.

@laviro @sharvit @MariaAga might know better

MariaAga replied:
Vendor 8.X includes some breaking changes to diff modal so if you need to update vendor please include the fixes for the diff.
There is a PR to upgrade to 8 including the fixes: https://github.com/theforeman/foreman/pull/7963

kamils-iRonin commented:
[test foreman]

kamils-iRonin commented:
I recorded a short demo

https://user-images.githubusercontent.com/37402924/109229983-65ca8a80-77c4-11eb-8c39-ba36647a2930.mov

timogoebel commented:
@kamils-iRonin: Thanks for recording the video. I really like what can be seen there.
I noticed two things:
- I think we don't need the setting to enable/disable the statuses.
- The search by host status should use some human-friendly name, not the internal model name.

ares commented:
I agree with @timogoebel and would also add one small ask. Could you please change the page header to something like "Host Status Overview"? Please ping me once you rebase the PR, I'm happy to test, I think it's getting close to be ready for merge.

timogoebel commented on app/controllers/host_statuses_controller.rb:
```diff
@@ -0,0 +1,10 @@
+class HostStatusesController < ApplicationController
+  def welcome
```
Do we actually have a welcome page defined somewhere?

kamils-iRonin committed:
Fixes #31638 - Adds host status overview page

timogoebel commented on app/presenters/host_status_presenter.rb:
```diff
@@ -0,0 +1,134 @@
+class HostStatusPresenter
+  class Collection < Array
+    def search_for(*search_options)
+      # TODO
```
What is missing here?

kamils-iRonin replied:
I forgot to delete this comment. The API does not support search and returns all statuses.

kamils-iRonin commented:
> Could you please change the page header to something like "Host Status Overview"? Please ping me once you rebase the PR, I'm happy to test, I think it's getting close to be ready for merge.

@ares it's ready to be tested :-)
![image](https://user-images.githubusercontent.com/37402924/114474952-94c38100-9bf7-11eb-8c23-8fb4e35964fd.png)

kamils-iRonin commented:
[test foreman]

timogoebel commented:
@ekohl: Do you have any idea, why the bot adds the WoC label all the time?

timogoebel commented:
@ezr-ondrej, @amirfefer: Do you guys mind taking a look here? I believe this is ready for a final review and testing.

ekohl commented:
I think that's the old bot which has a bit stricter requirements on the commit message but we've removed the reporting of that. I suspect it's [this line](https://github.com/theforeman/prprocessor/blob/f9b4ca9138957095e9b824181a1d5ef6bceee385/app.rb#L85) that triggers. In particular [line length of the first line](https://github.com/theforeman/prprocessor/blob/f9b4ca9138957095e9b824181a1d5ef6bceee385/github/pull_request.rb#L88).

ezr-ondrej reviewed (commented):
I think `Adds host status overview page` would be better commit message, the move from wreckingball can be in following commit lines.

@amirfefer could you have a look, I think that for #8285 we could reuse the controller. Maybe some parts of the components could be reused.

ares commented:
I've retested and the page looks really great. Two comments

1) could the page get a title, e.g. `Host Status Overview` (@amirfefer might know how to do it, I think this was made possible by https://github.com/theforeman/foreman/pull/7505)

2) in my case, I have 2 statuses (one from plugin, one from core) that seems to have "unknown" status, when I checked the host configuration status it was "out of sync", do you know what it means? Shouldn't that be under warning? Also why it says "Status 0"? Screenshot attached

![statuses](https://user-images.githubusercontent.com/109773/117622550-01885780-b173-11eb-8621-4eddb477f3e8.png)

@amirfefer could you please take a look at the code? That's something I can't review :-)

kamils-iRonin commented:
> 2\. in my case, I have 2 statuses (one from plugin, one from core) that seems to have "unknown" status, when I checked the host configuration status it was "out of sync", do you know what it means? Shouldn't that be under warning? Also why it says "Status 0"? Screenshot attached

@ares To correctly recognize and display statuses we need to set some variables in the models, like here [1].
Otherwise the status is marked as unknown and the label is `"Status #{value}"`. I'll take care of the core statuses, the configuration status seems to be left out.

[1] https://github.com/kamils-iRonin/foreman/blob/31638-Move-the-host-status-page-from-foreman_wreckingball-to-core/app/models/host_status/build_status.rb#L8-L17

ares commented:
If I understand correctly, you need a full list which may be hard for the configuration status (the number represents how many resources changed, failed etc). I think the lists you introduce can work, but we should also change the `to_global` implementation to rely on them, otherwise we duplicate the logic on two places. It will be a bit harder in configuration status, but the arrays like OK_STATUSES could be a range calculated by ConfigReportStatusCalculator. I think it's fine to proceed once the configuration status is resolved (the hardest one), we can then take care of the plugins.

ares commented:
Oh and in case there's all hosts unknown (e.g. configuration atm), should the overall status icon still be green?

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/Details.js:
```diff
@@ -0,0 +1,67 @@
+const Details = ({ data }) => {
+  const dataClone = useMemo(() => cloneDeep(data), [data]);
+
+  const columns = ['', 'Total', 'Owned'];
```
needs translation?

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/Details.js:
```diff
@@ -0,0 +1,67 @@
+import LinkOrLabel from './LinkOrLabel';
+
+const Details = ({ data }) => {
+  const dataClone = useMemo(() => cloneDeep(data), [data]);
```
No need to use a  deep clone nor `useMemo` here,
the given data from the API is immutable, the Table component needs mutable data, 
You can pass the data prop as mutable by using [asMutable](https://github.com/rtfeldman/seamless-immutable#asmutable)

kamils-iRonin replied:
nice! thanks 👍

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/GlobalStatusIcon.js:
```diff
@@ -0,0 +1,54 @@
+import CheckCircleIcon from '@patternfly/react-icons/dist/js/icons/check-circle-icon';
+import ExclamationTriangleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-triangle-icon';
+import ExclamationCircleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
+import QuestionCircleIcon from '@patternfly/react-icons/dist/js/icons/question-circle-icon';
```
This will create an issue during build, please use 
```js
 import { CheckCircleIcon, ExclamationTriangleIcon ,... } from '@patternfly/react-icons'
```

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/index.js:
```diff
@@ -0,0 +1,37 @@
+          <Fragment>
+            <StatusSkeleton />
+            <StatusSkeleton />
+          </Fragment>
```
There's a [SkeletonLoader](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/SkeletonLoader/index.js) component for this case

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/index.js:
```diff
@@ -0,0 +1,203 @@
+  error_owned_path: errorOwnedPath,
+  details,
+}) => {
+  const globalStatus = useMemo(
+    () =>
+      Math.max(
+        ...details
+          .filter(({ total }) => total > 0)
+          .map(({ global_status: gs }) => gs),
+        0
+      ),
+    [details]
+  );
```
data manipulation should be done via selectors, an example that demonstrates selectors usage:
```js
// StatusSelectors.js
import { selectAPIResponse } from '/API/APISelectors';

export const selectManipulatedStatus = (state, key) =>
   selectAPIResponse(state, key)
    .filter(({ total }) => total > 0)
    .map(({ global_status: gs }) => gs)
}

// index.js
import { selectManipulatedStatus } from './StatusSelectors'
  const {
     status,
     key,
   } = useAPI('get', url});
  const data =  useSelector((state, key) =>  selectManipulatedStatus(state, key)
  // return <Status> with calculated data
```

In some cases, like the useMemo, we can memorize selectors by [reselect](https://github.com/reduxjs/reselect), but we should consider using optimiztions only if we encounter performances issues.

This concern related also to data manipulation down below.

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/index.js:
```diff
@@ -0,0 +1,203 @@
+            'aria-expanded': expanded,
+          }}
+        >
+          <Grid style={{ width: '100%' }} hasGutter>
```
please use css classes instead of inline styling

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/index.js:
```diff
@@ -0,0 +1,203 @@
+          {details.length > 0 ? (
+            <Details data={details} />
+          ) : (
+            __('Nothing to show')
```
how about render `emptyState` component?

kamils-iRonin replied:
I would stick with just text, this component doesn't look good here

amirfefer reviewed (changes requested):
Thanks @kamils-iRonin, great feature 👍 

I've added some inline comments, repeated issues I've found:
- `useMemo` and other memorization utils should be used only if optimization is needed due to performance issues.
- Please use selectors for redux's data  manipulations (I've added an example)
- Please prefer using css classes over inline styling

kamils-iRonin commented on webpack/assets/javascripts/react_app/Root/apollo.js:
```diff
@@ -1,5 +1,8 @@
 import { ApolloClient, ApolloLink, InMemoryCache, from } from '@apollo/client';
+
+/* eslint-disable import/no-extraneous-dependencies */
```
no idea why this is needed 🤔

amirfefer replied:
does this lint issue occur in your local env?
can you delete it so we can make sure if it happens during tests?

kamils-iRonin commented:
@amirfefer Thank you for your review. I believe I addressed your comments.



> I think it's fine to proceed once the configuration status is resolved (the hardest one)

@ares I've created a new issue for configuration status [1] and I'm working on it. I'll open a separate PR so that it doesn't block this one. 



> Oh and in case there's all hosts unknown (e.g. configuration atm), should the overall status icon still be green?

I think it's fine, the green icon means no errors/warnings detected.

[1] https://projects.theforeman.org/issues/32670

ares commented:
@amirfefer could you please take a look at the new version? if the code is ready, I'm happy to re-test

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/LinkOrLabel.js:
```diff
@@ -0,0 +1,20 @@
+  if (path) {
+    return <a href={path}>{label}</a>;
+  }
+  return <Fragment>{label}</Fragment>;
```
nit: Fragments in react created for grouping and render mulitple components, is there a reason for wrapping `label` with a fragment?

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/Status/index.js:
```diff
@@ -0,0 +1,179 @@
+                path={warnTotalPath}
+                label={sprintf(__('Total: %s'), warnTotalCount)}
+              />
+              <br />
```
`<br>` tag purpose is to create a line break in a block of text, for elements margin please use css.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/SkeletonLoader/index.js:
```diff
@@ -7,14 +7,15 @@ import { translate as __ } from '../../../../react_app/common/I18n';
 
 const SkeletonLoader = ({
   status,
+  skeleton,
```
Makes sense
can you change it to `customSkeleton`?

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/HostStatusesSelectors.js:
```diff
@@ -0,0 +1,72 @@
+  GLOBAL_STATUS_ERROR,
+} from './constants';
+
+export const selectManipulatedStatuses = (state, key) => {
```
Selectors work the best in a modular structure, so we can reuse them in other places, plus it easier to read, maintain (optimize), and test: 
```js
// can be a predefined key
import { HOST_STATUS_KEY } from './consts'; 

export const selectHostStatuses = state => {
  const response = selectAPIResponse(state, HOST_STATUS_KEY);
}

export const selectHostStatusesDetails = state => {
   selectHostStatuses(state)?.results?.details
}

export const selectGlobalStatus = state =>  {
  const withTotal = 
    selectHostStatusesDetails(state)
      .filter(({ total }) => total > 0)
      .map(({ global_status: gs }) => gs),
    return Math.max(withTotal, 0);
}

export const selectOKStatueses = state => {
// ...
} 

// etc
```

amirfefer reviewed (changes requested):
Thanks @kamils-iRonin 👍 
I left a few inline comments.

Please add some tests.

amirfefer commented on webpack/assets/javascripts/react_app/common/hooks/API/APIHooks.js:
```diff
@@ -18,6 +18,7 @@ import { APIActions } from '../../../redux/API';
 export const useAPI = (method, url, options) => {
   const dispatch = useDispatch();
   const keyRef = useRef(options?.key);
+  const optionsRef = useRef(options);
```
Is there a reason for creating a ref for `options`?
Refs are used for storing mutable values, similar to instance variables in a class, but the `options` object here shouldn't be changed, no?

https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables

kamils-iRonin replied:
this is to solve the infinite loop of `useEffect` when options are passed, it took me a while to figure out why the request is stuck on `PENDING` when `options` with custom key are passed
I also don't see `useAPI` to be used with `options` anywhere in the existing codebase

amirfefer replied:
`useEffect` compares the values in the dependency array by reference (shallow), which causes the infinite loop.
in your case, `options` object is  constant, so just define it in the `constants.js` file:
```js
export const API_OPTIONS = { key: 'some-key'}
```
this will stop your infinite loop.

if options object is dynamic and based on other props, wrapping it with `useMemo` works:
```js
const options = useMemo(() => optionCreator(prop), [prop]);
```

amirfefer commented on webpack/assets/javascripts/react_app/components/HostStatuses/HostStatusesSelectors.js:
```diff
@@ -0,0 +1,74 @@
+  selectHostStatus(state, statusName)?.description || '';
+
+/* eslint-disable camelcase */
+export const selectHostStatusOkTotalPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.ok_total_path;
+
+export const selectHostStatusOkOwnedPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.ok_owned_path;
+
+export const selectHostStatusWarnTotalPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.warn_total_path;
+
+export const selectHostStatusWarnOwnedPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.warn_owned_path;
+
+export const selectHostStatusErrorTotalPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.error_owned_path;
+
+export const selectHostStatusErrorOwnedPath = (state, statusName) =>
+  selectHostStatus(state, statusName)?.error_total_path;
```
How about some aggregation?

```js
export const selectHostStatusTotalPath = () => ({
  okTotalPath: selectHostStatus(state, statusName)?.ok_total_path,
  warbTotalPath: selectHostStatus(state, statusName)?.warn_total_path,
  // ...etc
})

export const selectHostStatusOwnedPath = () => ({
  okOwnedlPath: selectHostStatus(state, statusName)?.ok_owned_path;,
  warnOwnedPath: sselectHostStatus(state, statusName)?.warn_owned_path;,
  // ...etc
})
```
This  aggregates the data and saves some redundant imports later

amirfefer reviewed (dismissed):
Thanks @kamils-iRonin, Overall looks good 👍 
JS testing is missing.

kamils-iRonin commented:
@amirfefer Thanks for the review, I addressed your comments and added some tests

ezr-ondrej commented on app/registries/foreman/access_permissions.rb:
```diff
@@ -312,6 +312,10 @@
   end
 
+  permission_set.security_block :host_statuses do |map|
+    map.permission :view_statuses, {:"api/v2/host_statuses" => [:index]}
```
do we really need another permission for this? Isn't view_hosts the permission to allow this?

kamils-iRonin replied:
it would make sense to use `view_hosts` permission 👍   just wondering, do we want to be able to grant permissions only to particular statuses?

tbrisker replied:
I would rather start without a specific permission and only add it if it is needed in the future, additional permissions add complexity and reduce performance.

ezr-ondrej commented on app/views/api/v2/host_statuses/index.json.rabl:
```diff
@@ -0,0 +1,17 @@
+node :details do |presenter|
+  presenter.all_statuses.map do |status|
+    {
+      label: presenter.labels.fetch(status, "Status #{status}"),
```
`status` is a number here, it is not very informative to the user, could we translate it to something more meaningful?

kamils-iRonin replied:
status number is only displayed if labels are not defined, see build status labels

https://github.com/kamils-iRonin/foreman/blob/d410d067f449ec2102c4a2fec9d5431d0f05c8c6/app/models/host_status/build_status.rb#L12-L17

timogoebel replied:
Yep, I'm wondering if we want to simply hide these statuses until the labels are defined in the respective plugins.

kamils-iRonin replied:
easy to do let's hide statuses if `LABEL` variable is not defined. I think we should leave some generic label like `Missing label for status X` in case a new status option has been introduced but no label has been added

ezr-ondrej reviewed (commented):
Works well, just two finds, I believe the React code is polished enough already :)

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -148,6 +149,7 @@ const coreComponets = [
   { name: 'PersonalAccessTokens', type: PersonalAccessTokens },
   { name: 'ClipboardCopy', type: ClipboardCopy },
   { name: 'LabelIcon', type: LabelIcon },
+  { name: 'HostStatuses', type: HostStatuses },
```
I don't see us mounting it anywhere from Rails, I believe that means we don't need to expose it in componentRegistry.
Or am I missing something?

kamils-iRonin replied:
you're right, fixed

ezr-ondrej reviewed (approved):
Great addition :+1:

timogoebel commented:
Glad to see this merged. For reference: The follow up PR is in #8566
